kythe/go/serving/pipeline/beamio/leveldb.go

```diff
@@ -0,0 +1,380 @@
+	opts := &opt.Options{
+		BlockSize: 5 * opt.MiB,
+		Comparer:  keyComparer{},
+	}
+
+	var totalElements int
+	defer func(start time.Time) {
+		log.Printf("Shard %04d: %s (size: %d)", shard, time.Since(start), totalElements)
+	}(time.Now())
+	md := tableMetadata{Shard: shard + 1}
+
+	var els []keyValue
+	var kv keyValue
+	for e(&kv) {
+		els = append(els, kv)
+	}
+	sort.Slice(els, func(i, j int) bool {
+		return bytes.Compare(els[i].Key, els[j].Key) < 0
+	})
+
+	// Remove duplicate keys
+	j := 1
+	for i := 1; i < len(els); i++ {
+		if bytes.Equal(els[j-1].Key, els[i].Key) {
+			duplicateLevelDBKeysCounter.Inc(ctx, 1)
+		} else {
+			els[j] = els[i]
+			j++
+		}
+	}
+	els = els[:j]
+
+	// Encode keys for LevelDB
+	for i := 0; i < len(els); i++ {
+		md.Seq++
+		els[i].Key = makeLevelDBKey(uint64(md.Seq), els[i].Key)
+	}
+
+	totalElements = len(els)
+	md.First = els[0].Key
+	md.Last = els[len(els)-1].Key
+
+	// Write each sorted key-value to an SSTable.
+	f, err := openWrite(ctx, filepath.Join(w.Path, fmt.Sprintf("%06d.ldb", md.Shard)))
+	if err != nil {
+		return err
+	}
+	wr := table.NewWriter(f, opts)
+	for _, kv := range els {
+		if err := wr.Append(kv.Key, kv.Value); err != nil {
+			return err
+		}
+	}
+	if err := wr.Close(); err != nil {
+		return err
+	} else if err := f.Close(); err != nil {
+		return err
+	}
+	md.Size = wr.BytesLen()
+
+	emit(md)
+	return nil
+}
+
+type shardKeyValue struct{ Shards int }
+
+func (s *shardKeyValue) ProcessElement(kv keyValue) (int, keyValue) {
+	h := crc32.NewIEEE()
+	h.Write(kv.Key)
+	return int(h.Sum32()) % s.Shards, kv
+}
+
+type encodeKeyValue struct {
+	Coder beam.EncodedCoder
+}
+
+func (e *encodeKeyValue) ProcessElement(key beam.T, val beam.U) (keyValue, error) {
+	c := beam.UnwrapCoder(e.Coder.Coder)
+	keyEnc := exec.MakeElementEncoder(c.Components[0])
+	var keyBuf bytes.Buffer
+	if err := keyEnc.Encode(exec.FullValue{Elm: key}, &keyBuf); err != nil {
+		return keyValue{}, err
+	} else if _, err := binary.ReadUvarint(&keyBuf); err != nil {
+		return keyValue{}, fmt.Errorf("error removing varint prefix from key encoding: %v", err)
+	}
+	valEnc := exec.MakeElementEncoder(c.Components[1])
+	var valBuf bytes.Buffer
+	if err := valEnc.Encode(exec.FullValue{Elm: val}, &valBuf); err != nil {
+		return keyValue{}, err
+	} else if _, err := binary.ReadUvarint(&valBuf); err != nil {
+		return keyValue{}, fmt.Errorf("error removing varint prefix from value encoding: %v", err)
+	}
+	return keyValue{Key: keyBuf.Bytes(), Value: valBuf.Bytes()}, nil
+}
+
+// A keyValue is a concrete form of a Beam KV.
+type keyValue struct {
+	Key   []byte `json:"k"`
+	Value []byte `json:"v"`
+}
+
+// makeLevelDBKey constructs an internal LevelDB key from a user key.  seq is
+// the sequence number for the key-value entry within the LevelDB.
+func makeLevelDBKey(seq uint64, key []byte) []byte {
+	const typ = 1 // value (vs. deletion)
+	k := make([]byte, len(key)+8)
```

Comment:
you've got a few different places where you're adding or subtracting 8, or shifting by 8.  What does the 8 represent?  Please consider creating a constant for this (assuming it's all the same thing), and/or a method or methods to do the offsetting.

Reply:
There are two "types" of entries in a LevelDB SStable: a value update or a value deletion (1 vs. 0).  We only want to insert values so that's the constant.

Reply:
✓